lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -405,6 +412,313 @@ bool ADContext::supportsFloatingPointDifferentiation(Type type) const {
+/// derivative `v̇` at this place is certainly null. Conversely when variable `v`
+/// is not useful, then whatever the value of `v̇`, this value does not matter
+/// for the final result. Symmetric reasoning applies for the reverse mode of
+/// AD: observing that differentiated variables go upstream, we see that a
+/// useless variable has a null derivative, in other words the partial
+/// derivative of the output with respect to this variable is null. Conversely
+/// when variable `v` is not varied, then whatever the value of `v`, this value
+/// does not matter for the final result.
+///
+/// Reference:
+/// Laurent Hascoët. Automatic Differentiation by Program Transformation. 2017.
+
+namespace {
+
+class DifferentiableActivityInfo;
+
+class DifferentiableActivityAnalysis
+  : public FunctionAnalysisBase<DifferentiableActivityInfo> {
+private:
+  DominanceAnalysis *dominanceAnalysis = nullptr;
+
+public:
+  explicit DifferentiableActivityAnalysis()
+    : FunctionAnalysisBase(AnalysisKind::DifferentiableActivity) {}
+
+  static bool classof(const SILAnalysis *s) {
+    return s->getKind() == AnalysisKind::DifferentiableActivity;
+  }
+
+  virtual bool shouldInvalidate(SILAnalysis::InvalidationKind k) override {
+    return k & InvalidationKind::Everything;
+  }
+
+  virtual
+  DifferentiableActivityInfo *newFunctionAnalysis(SILFunction *f) override;
+
+  virtual void initialize(SILPassManager *pm) override;
+};
+
+class DifferentiableActivityInfo {
+private:
+  SILFunction &function;
+
+  /// Input values, i.e. parameters (both direct and indirect).
+  SmallVector<SILValue, 4> inputValues;
+  /// Output values, i.e. individual values (not the final tuple) being returned
+  /// by the `return` instruction.
+  SmallVector<SILValue, 4> outputValues;
+
+  /// The set of useful variables, indexed by the corresponding dependent value
+  /// (output) index.
+  SmallVector<SmallDenseSet<SILValue>, 4> usefulValueSets;
+  /// The set of useful variables, indexed by the corresponding independent
+  /// value (input) index.
+  SmallVector<SmallDenseSet<SILValue>, 4> variedValueSets;
+
+  /// Perform analysis and populate sets.
+  void analyze();
+
+public:
+  explicit DifferentiableActivityInfo(SILFunction &f);
+
+  bool isIndependent(SILValue value,
+                     const SILReverseAutoDiffIndices &indices) const;
+  bool isDependent(SILValue value,
+                   const SILReverseAutoDiffIndices &indices) const;
+  bool isVaried(SILValue value,
+                unsigned independentVariableIndex) const;
+  bool isUseful(SILValue value,
+                unsigned dependentVariableIndex) const;
+  bool isVaried(SILValue value,
+                llvm::SmallBitVector parameterIndices) const;
+  bool isActive(SILValue value,
+                const SILReverseAutoDiffIndices &indices) const;
+};
+
+} // end anonymous namespace
+
+DifferentiableActivityInfo *
+DifferentiableActivityAnalysis::newFunctionAnalysis(SILFunction *f) {
```

Comment:
"create" would be more standard with llvm conventions, "new" is a cocoa convention.

Reply:
I was following `DominanceAnalysis`:

https://github.com/apple/swift/blob/67cf94fb5517c16bcb3ffe1c0f5dbc10351553d0/include/swift/SILOptimizer/Analysis/DominanceAnalysis.h#L73

Shall I change that upstream too?

Reply:
And, it's a virtual method on `swift::FunctionAnalysisBase`.